sqldelight-runtime/src/main/java/com/squareup/sqldelight/SqlDelightDatabase.kt

```diff
@@ -1,93 +1,81 @@
-) : Transactor, Closeable {
-
-  private val transactions = ThreadLocal<SqliteTransaction>()
-
-  private val transaction = object : Transaction {
-    override fun markSuccessful() {
-      helper.writableDatabase.setTransactionSuccessful()
-    }
-
-    override fun yieldIfContendedSafely(): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely()
-    }
-
-    override fun yieldIfContendedSafely(sleepAmount: Long, sleepUnit: TimeUnit): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely(sleepUnit.toMillis(sleepAmount))
-    }
-
-    override fun end() {
-      val transaction = transactions.get() ?: throw IllegalStateException("Not in transaction.")
-      transactions.set(transaction.parent)
-      helper.writableDatabase.endTransaction()
-      if (transaction.commit) {
-        transaction.forEach { function -> afterTransaction(function) }
-      }
-    }
-
-    override fun close() {
-      end()
-    }
-  }
-
-  override fun newTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListener(transaction)
-
-    return this.transaction
-  }
-
-  override fun newNonExclusiveTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListenerNonExclusive(transaction)
+  internal val helper: SqlDatabaseHelper
+) {
+  private val transactions = ThreadLocal<Transaction>()
+  private val actions = ArrayList<() -> Unit>()
 
-    return this.transaction
-  }
-
-  override fun afterTransaction(function: () -> Unit) {
+  fun afterTransaction(function: () -> Unit) {
     val transaction = transactions.get()
     if (transaction != null) {
-      transaction.add(function)
+      actions.add(function)
     } else {
       function()
     }
   }
 
-  override fun close() {
+  fun close() {
     helper.close()
   }
-}
 
-private class SqliteTransaction(
-    val parent: SqliteTransaction?
-) : LinkedHashSet<() -> Unit>(), SQLiteTransactionListener {
-  var commit: Boolean = false
+  fun transaction(body: Transaction.() -> Unit) {
+    val parent = transactions.get()
+    val transaction = Transaction()
 
-  override fun onBegin() {}
+    if (parent == null) {
+      helper.getConnection().beginTransaction()
+    }
 
-  override fun onCommit() {
-    commit = true
+    try {
+      transactions.set(transaction)
+      transaction.body()
+      transaction.successful = true
+    } catch(e: RollbackException) {
+      if (parent != null) throw e
+    } finally {
+      transactions.set(parent)
+      if (parent == null) {
+        if (!transaction.successful || !transaction.childrenSuccessful) {
+          helper.getConnection().rollbackTransaction()
+        } else {
+          helper.getConnection().commitTransaction()
+          while (actions.isNotEmpty()) actions.removeAt(0).invoke()
```

Comment:
I find it super weird that this type is called Database and it’s being mutated on a single transaction execution. I think this class should be called Transaction?

Reply:
unless I’m misunderstanding the scope of these actions

Reply:
no they're strictly transaction-tied. I like `Transaction`. It gets subclassed by the generated database type anyway but it better reflects the class